review

pedruuv opened the pull request.

pedruuv committed:
review

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/controller/DeliverController.java:
```diff
@@ -0,0 +1,39 @@
+@RequestMapping("/delivery")
+@RequiredArgsConstructor
+public class DeliverController {
+    @Autowired
```
If you are using @RequiredArgsConstructor it means Lombok will create constructor for each service you have as a parameter in your class. And If you are using @Autowired - you provide dependency injection using Spring DI. Yu you have to you only one of this cases but not both

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/controller/DeliverController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @PostMapping
+    public ResponseEntity<Deliver> newDelivery(@Valid @RequestBody NewDeliverDto newDeliver){
+        Deliver deliver = new Deliver(newDeliver.productName(), newDeliver.clientName(), newDeliver.clientEmail(), newDeliver.clientPhone(), newDeliver.distance());
```
Controller should not contains any business logic. This must be replaced into service

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/model/LocationUpdate.java:
```diff
@@ -0,0 +1,21 @@
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
```
seems you don't need it here @AllArgsConstructor

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/model/LocationUpdate.java:
```diff
@@ -0,0 +1,21 @@
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class LocationUpdate {
+    @Id
+    private UUID id;
```
if it's table so probably you need to have @GeneratedValue(strategy = GenerationType.UUID)

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/DeliveryStatusUpdater.java:
```diff
@@ -0,0 +1,59 @@
+
+@Service
+@RequiredArgsConstructor
+public class DeliveryStatusUpdater {
```
Usually, in real projects class will newer be injected, and will be injected his interface instead. There are lot of reasons for that. So please create interface for this class, and every time you need this class just use his interface so you will avoid hard cohesion between services

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/DeliveryStatusUpdater.java:
```diff
@@ -0,0 +1,59 @@
+            put(percentage -> percentage == DELIVERED_PERCENTAGE, DeliveryStatus.DELIVERED);
+        }};
+
+        return statusMap.entrySet().stream().filter(entry -> entry.getKey().test(percentageRemaining))
```
This logic seems incorrect... For example If percentageRemaining=0.1 and we are expecting OUT_FOR_DELIVERY_PERCENTAGE but we will get AWAITING_PICKUP_PERCENTAGE  because predicate will match with  percentageRemaining <= AWAITING_PICKUP_PERCENTAGE.

pedruuv replied:
> This logic seems incorrect... For example If percentageRemaining=0.1 and we are expecting OUT_FOR_DELIVERY_PERCENTAGE but we will get AWAITING_PICKUP_PERCENTAGE because predicate will match with percentageRemaining <= AWAITING_PICKUP_PERCENTAGE.

Actually, this doesn't happen, because even if percentageRemaining <= AWAITING_PICKUP_PERCENTAGE is true, there is a second condition to be evaluated, which is percentageRemaining > PICKUP_PERCENTAGE, which would be false, and therefore the predicate for percentageRemaining = 0.1 would not be satisfied, and the check would go on to the next conditions.

yuraf640 replied:
Ahh, my bad,  i missed the second condition, so you are right!

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/DeliveryStatusUpdater.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    private DeliveryStatus getNewStatusBasedOnPercentage(double percentageRemaining, DeliveryStatus currentStatus) {
+        Map<Predicate<Double>, DeliveryStatus> statusMap = new LinkedHashMap<>(){{
```
this Map must be generated only once in constructor to not generate it every time we call getNewStatusBasedOnPercentage()

yuraf640 commented on delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/KafkaConsumer.java:
```diff
@@ -0,0 +1,30 @@
+    public void consumeLocationUpdate(String trackingUpdate) {
+        try {
+            LocationUpdate locationUpdate = mapper.readValue(trackingUpdate, LocationUpdate.class);
+            Deliver deliver = repository.findById(locationUpdate.getId()).orElseThrow();
```
orElseThrow should throw something sensible like NotFoundException

yuraf640 commented on notification-service/src/main/java/com/tracking_delivery_system/notification_service/service/KafkaConsumer.java:
```diff
@@ -0,0 +1,66 @@
+    private synchronized void checkAndPrintMessage(String deliveryId) {
+        Map<String, Object> deliveryMap = deliveriesMap.get(deliveryId);
+
+        if (deliveryMap.containsKey("location-update") && deliveryMap.containsKey("delivery-status")) {
```
i would recommend you divide this logic to have separate "if" for location-update and delivery-status because it is separate topics and and as i seen they are not combined. Or maybe i missed something - please correct me

pedruuv replied:
I used only one "if" for both topics, because inside this "if" I’m printing the messages from both topics in a single message. So this "if" ensures that the message will only be shown in the console when there is data from both topics.

pedruuv replied:
> i would recommend you divide this logic to have separate "if" for location-update and delivery-status because it is separate topics and and as i seen they are not combined. Or maybe i missed something - please correct me

I used only one "if" for both topics, because inside this "if" I’m printing the messages from both topics in a single message. So this "if" ensures that the message will only be shown in the console when there is data from both topics.

yuraf640 commented on tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/service/DeliveryProgressSimulator.java:
```diff
@@ -0,0 +1,54 @@
+    private Random random = new Random();
+
+    public void simulateDeliverProgress(LocationUpdate locationUpdate) {
+        final long[] distance = {locationUpdate.getDistance()};
```
why you create array from string it in all subsequent steps you are always using only first element of this array (distance[0])?

pedruuv replied:
To allow the variable distance to be modified within the lambda expression.

pedruuv replied:
> why you create array from string it in all subsequent steps you are always using only first element of this array (distance[0])?

To allow the variable distance to be modified within the lambda expression.

yuraf640 replied:
Interesting solution... But i think you could avoid this redundant complexity just having one variable inside of lambda "if" condition, besides you don't modify distance for lambda's "else" condition. So it's up to you, i don't have strict position against it

yuraf640 reviewed (changes requested):
In general looks good, but please address my comments

pedruuv committed:
I've corrected everything in the comments and implemented a geolocation feature

# Conflicts:
#	delivery-status-service/Dockerfile
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/config/KafkaTopicConfig.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/controller/DeliverController.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/dto/NewDeliverDto.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/model/Deliver.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/model/DeliveryStatus.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/model/LocationUpdate.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/model/StatusUpdate.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/repository/DeliverRepository.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/DeliveryStatusUpdater.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/KafkaConsumer.java
#	delivery-status-service/src/main/java/com/tracking_delivery_system/delivery_status_service/service/KafkaProducer.java
#	docker-compose.yml
#	notification-service/src/main/java/com/tracking_delivery_system/notification_service/model/LocationUpdate.java
#	notification-service/src/main/java/com/tracking_delivery_system/notification_service/model/StatusUpdate.java
#	notification-service/src/main/java/com/tracking_delivery_system/notification_service/service/KafkaConsumer.java
#	tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/config/KafkaTopicConfig.java
#	tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/model/LocationUpdate.java
#	tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/repository/LocationUpdateRepository.java
#	tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/service/DeliveryProgressSimulator.java
#	tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/service/KafkaConsumer.java
#	tracking-service/src/main/java/com/tracking_delivery_system/tracking_service/service/KafkaProducer.java

yuraf640 commented on geolocation_service/src/main/java/com/tracking_delivery_system/geolocation_service/service/CoordinatesService.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    public double getTotalDistance(Coordinates origin, Coordinates destination) {
+        String url = String.format("https://router.project-osrm.org/route/v1/driving/%s,%s;%s,%s",
```
It's better to have URLs in configurations, so later it will be easy to switch to another service without code change

yuraf640 commented on geolocation_service/src/test/java/com/tracking_delivery_system/geolocation_service/GeolocationServiceApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+class GeolocationServiceApplicationTests {
+
+	@Test
+	void contextLoads() {
```
seems this test is useless

yuraf640 reviewed (approved):
Very good! Please switch to unit tests
